apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -514,6 +569,169 @@ fn parse_url(value: &Node<Value>, coordinate: &str, sources: &SourceMap) -> Resu
     Ok(url)
 }
 
+fn validate_header_arg<'a>(
+    directive_name: &Name,
+    argument_name: &str,
+    headers: &'a Node<Value>,
+    source_map: &SourceMap,
+    object: Option<&Name>,
+    field: Option<&Name>,
+) -> Result<&'a Node<Value>, Vec<Message>> {
+    let mut errors = Vec::new();
+
+    headers
+        .as_list()
+        .map(|l| l.iter().filter_map(|o| o.as_object()).collect_vec())
+        .unwrap_or_else(|| headers.as_object().map(|o| vec![o]).unwrap_or_default())
+        .iter()
+        .for_each(|arg_pairs| {
+            let mut has_as_or_value = false;
+            let mut unique_header_set = HashSet::new();
+
+            arg_pairs.iter().for_each(|(key, value)| {
```

Comment:
I think rather than trying to keep state in `has_as_or_value`, I would `.find()` each of the relevant arguments and then do the checks. Might be easier to read.

Something like this:

```rust
let name_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("name")).then_some(value));
let as_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("as")).then_some(value));
let value_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("value")).then_some(value));
```

Reply:
i like this suggestion, and i think it makes the as/value rule easier to write too 
```rs
if let (Some(as_arg), Some(value_arg)) = (as_arg, value_arg) {
  errors.push(...);
}
```